fdbserver/DataDistribution.actor.cpp

```diff
@@ -3519,7 +3585,78 @@ ACTOR Future<Void> monitorStorageServerRecruitment(DDTeamCollection* self) {
 	}
 }
 
+ACTOR Future<Void> checkAndRemoveInvalidLocalityAddr(DDTeamCollection* self) {
+	state double start = now();
+	state Transaction tr(self->cx);
+
+	loop {
+		try {
+			wait(delay(SERVER_KNOBS->DD_CHECK_INVALID_LOCALITY_DELAY));
+
+			tr.setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::READ_LOCK_AWARE);
+
+			// Because worker's processId can be changed when its locality is changed, we cannot watch on the old
+			// processId; This actor is inactive most time, so iterating all workers incurs little performance overhead.
+			state vector<ProcessData> workers = wait(getWorkers(&tr));
+			state std::set<AddressExclusion> existingAddrs;
+			for (int i = 0; i < workers.size(); i++) {
+				const ProcessData& workerData = workers[i];
+				AddressExclusion addr(workerData.address.ip, workerData.address.port);
+				existingAddrs.insert(addr);
+				if (self->invalidLocalityAddr.count(addr) &&
+				    self->isValidLocality(self->configuration.storagePolicy, workerData.locality)) {
+					// The locality info on the addr has been corrected
+					self->invalidLocalityAddr.erase(addr);
+				}
+			}
+
+			wait(yield(TaskPriority::DataDistribution));
```

Comment:
I checked the other places that uses `getWorkers()`.  I found that they all assume it is fast to iterate through the list of workers. 
Checking `isValidLocality ` does increase the time spent on one worker, but I'm not sure how long it takes. 
If we want to add `wait` in the for-loop-of-workers, we need to know the max number of workers to iterate without wait, which is unclear. So I didn't add `wait` inside the loop.

Reply:
Ok, I think that's fine for now.